src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,418 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Linq;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    /// <summary>
+    /// An endpoint used to acquire the diagnostics stream for a runtime instance.
+    /// </summary>
+    internal interface IIpcEndpoint
+    {
+        /// <summary>
+        /// Wait for an available diagnostics connection to the runtime instance.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>
+        /// A task the completes when a diagnostics connection to the runtime instance becomes available.
+        /// </returns>
+        Task WaitForConnectionAsync(CancellationToken token);
+
+        /// <summary>
+        /// Connects to the underlying IPC Transport and opens a read/write-able Stream
+        /// </summary>
+        /// <returns>A Stream for writing and reading data to and from the target .NET process</returns>
+        Stream Connect();
+    }
+
+    internal abstract class BaseIpcEndpoint : IIpcEndpoint
+    {
+        private readonly Queue<StreamTarget> _targets = new Queue<StreamTarget>();
+
+        private bool _disposed;
+        private Stream _stream;
+
+        /// <inheritdoc cref="IIpcEndpoint.Connect"/>
+        /// <remarks>
+        /// This will block until the diagnostic stream is provided. This block can happen if
+        /// the stream is acquired previously and the runtime instance has not yet reconnected
+        /// to the reversed diagnostics server.
+        /// </remarks>
+        public Stream Connect()
+        {
+            using var target = new StreamEventTarget();
+
+            RegisterTarget(target);
+
+            target.Handle.WaitOne();
+
+            return target.Stream;
+        }
+
+        /// <inheritdoc cref="IIpcEndpoint.WaitForConnectionAsync(CancellationToken)"/>
+        public async Task WaitForConnectionAsync(CancellationToken token)
+        {
+            bool isConnected = false;
+            do
+            {
+                using (var target = new StreamTaskTarget(token))
+                {
+                    Stream stream = null;
+                    try
+                    {
+                        RegisterTarget(target);
+
+                        // Wait for a stream to be provided by the target
+                        stream = await target.Task;
```

Comment:
There's a race condition where if the WaitForConnectionAsync acquires the stream and the stream passes the test, that it can contend with providing the stream to other consumers with something else (like ServerIpcEndpoint.SetStream) also providing another stream at the same time. I should be able to solve this by using a semaphore that represents "allowed to mutate the stream field" so that WaitForConnection can decrement it while it is testing the stream and then increment it when providing it again. The other callers that can provide the stream will also do the same.